### PR TITLE
samples: nrf9160: modem_shell: AT command mode and pipelining

### PR DESCRIPTION
**Pull request by jhirsi**
This PR adds a separate plain AT command mode with AT command pipelining support to MoSh.
Additionally, as an enabler, a common workq implemented and taken into use within MoSh.

### CODE REVIEW

File: doc/nrf/releases/release-notes-changelog.rst [hunk 1]
```diff
@@ -148,6 +148,9 @@ nrf9160 Samples
 
       * An option ``--interval`` (in seconds) to neighbor cell measurements in continuous mode  (``link ncellmeas --continuous``).
         When using this option, a new measurement is started in each interval.
+      * Separate plain AT command mode that can be started by typing: ``at at_cmd_mode start``.
+        Kconfig options are available for configuring AT command termination options, default: CR termination.
+        In AT command mode, maximum of 10 AT commands can be pipelined by using ``|`` as a deliminator character between pipelined AT commands.
```
[the remark this answers is not in the document]
Comment (jhirsi) [reply doc/nrf/releases/release-notes-changelog.rst]: Thanks. Changed.

**Review by peknis — APPROVED** [review]

File: samples/nrf9160/modem_shell/src/at/at_cmd.c [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	int total = ARRAY_SIZE(mosh_at_cmd_list);
+
+	ret = at_params_list_init(&at_param_list, MOSH_AT_CMD_MAX_PARAM);
+	if (ret) {
+		printk("Could not init AT params list, error: %d\n", ret);
+		return ret;
+	}
+	ret = -ENOENT;
+	for (int i = 0; i < total; i++) {
+		if (at_cmd_mode_util_mosh_cmd_casecmp(at_cmd, mosh_at_cmd_list[i].string)) {
```
Comment (hiltunent) [inline samples/nrf9160/modem_shell/src/at/at_cmd.c:75]: I don't really understand the code, e.g. why is the at_cmd_mode_util_mosh_cmd_casecmp() needed? Some more comments could be helpful

Comment (hiltunent) [inline samples/nrf9160/modem_shell/src/at/at_cmd.c:75]: I don't really understand the code, e.g. why is the at_cmd_mode_util_mosh_cmd_casecmp() needed? Some more comments could be helpful

File: samples/nrf9160/modem_shell/src/at/at_cmd_mode.c [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+
+#define AT_CMD_MODE_AT_CMD_MAX_LEN CONFIG_SHELL_CMD_BUFF_SIZE
+#define AT_CMD_MODE_PIPELINED_AT_CMD_MAX_LEN 1023
+
+static uint8_t at_buf[AT_CMD_MODE_AT_CMD_MAX_LEN + 1];
+static uint8_t at_buf_pipelined[AT_CMD_MODE_PIPELINED_AT_CMD_MAX_LEN + 1];
+
+AT_MONITOR(mosh_at_cmd_mode_handler, ANY, at_cmd_mode_event_handler, PAUSED);
+
+
```
Comment (hiltunent) [inline samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:41]: extra line

Comment (hiltunent) [inline samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:41]: extra line

File: samples/nrf9160/modem_shell/src/at/at_cmd_mode.c [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+		at_cmd_mode_send_at_cmd_from_at_buf();
+	}
+	data->pipe_cnt = 0;
+}
+
+static void at_cmd_mode_cmd_rx_handler(uint8_t character)
+{
+	writing = true;
+
+	/* Handle control characters */
```
Comment (hiltunent) [inline samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:141]: at_cmd_mode_cmd_rx_handler() is very long function. Consider placing for example this part to a separate function.

Comment (hiltunent) [inline samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:141]: at_cmd_mode_cmd_rx_handler() is very long function. Consider placing for example control and terminal character setting to a separate function.

Comment (hiltunent) [inline samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:141]: at_cmd_mode_cmd_rx_handler() is very long function. Consider placing for example control and terminal character setting to a separate function.

File: samples/nrf9160/modem_shell/src/at/at_cmd_mode.c [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+	}
+
+	k_work_init(&at_cmd_mode_work.work, at_cmd_mode_worker);
+	at_cmd_mode_work.pipe_cnt = 0;
+
+	at_monitor_resume(mosh_at_cmd_mode_handler);
+
+	term_mode = CONFIG_MOSH_AT_CMD_MODE_TERMINATION;
+
+	at_cmd_mode_active = !at_cmd_mode_active;
```
Comment (hiltunent) [inline samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:323]: or simply set at_cmd_mode_active to true, easier to read

Comment (hiltunent) [inline samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:323]: or simply set at_cmd_mode_active to true, easier to read. Also, do you need the following if?

File: samples/nrf9160/modem_shell/src/at/at_cmd_mode.c [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+#if defined(CONFIG_MOSH_PING)
+		printk("  ICMP Ping: ");
+		printk("AT+NPING=<addr>[,<payload_length>,<timeout_msecs>,<count>"
+		       "[,<interval_msecs>[,<cid>]]]\n");
+#endif
+		printk("Other custom functionalities:\n");
+		printk("  AT command pipelining, for example:\n"
+		       "    at+cgmr|at+cfun?|at+nping=\"example.com\"\n");
+		printk("===========================================================\n");
+		at_cmd_mode_active = true;
```
Comment (hiltunent) [inline samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:336]: this line also don't make sense

Comment (hiltunent) [inline samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:336]: this line also don't make sense

**Review by hiltunent — CHANGES_REQUESTED** [review]

File: samples/nrf9160/modem_shell/src/at/at_cmd.c (hunk 2 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/at/at_cmd.c:75]: at_cmd_mode_util_mosh_cmd_casecmp() parses if given command is one of the mosh specific custom AT commands, ignoring the case. I will add more comments.

File: samples/nrf9160/modem_shell/src/at/at_cmd_mode.c (hunk 5 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:323]: yeah, need to simplify

File: samples/nrf9160/modem_shell/src/at/at_cmd_mode.c (hunk 4 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:141]: Agree that it is quite long, but in this case, IMO, it does not improve readability if splitting to subfunctions, mostly due to global variables that are used. For now, I'll left that as it is.

File: samples/nrf9160/modem_shell/src/at/at_cmd_mode.c (hunk 5 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:323]: Simplified. Thanks for the comments!

File: samples/nrf9160/modem_shell/src/at/at_cmd_mode.c (hunk 6 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:336]: Yes, removed.

**Review by hiltunent — APPROVED** [review]

File: samples/nrf9160/modem_shell/src/at/at_cmd.c [hunk 7]
```diff
@@ -72,7 +76,8 @@ int at_cmd_mosh_parse(const char *at_cmd)
 	}
 	ret = -ENOENT;
 	for (int i = 0; i < total; i++) {
-		if (at_cmd_mode_util_mosh_cmd_casecmp(at_cmd, mosh_at_cmd_list[i].string)) {
+		/* Check if given command is MoSh speific AT command */
```
Comment (tokangas) [inline samples/nrf9160/modem_shell/src/at/at_cmd.c:79]: Typo: "specific"

**Review by tokangas — APPROVED** [review]

File: samples/nrf9160/modem_shell/src/at/at_cmd_mode.c [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+	/* Echo a line feed */
+	printk("\n");
+
+	writing = false;
+
+	k_work_submit_to_queue(&mosh_common_work_q, &at_cmd_mode_work);
+}
+
+#define CHAR_CAN 0x18
+#define CHAR_DC1 0x11
```
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:265]: Yes, added.

File: samples/nrf9160/modem_shell/src/at/at_cmd_mode.c [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+		printk("  ICMP Ping: ");
+		printk("AT+NPING=<addr>[,<payload_length>,<timeout_msecs>,<count>"
+		       "[,<interval_msecs>[,<cid>]]]\n");
+#endif
+		printk("===========================================================\n");
+		at_cmd_mode_active = true;
+		at_cmd_mode_dont_print = true;
+	}
+
+	at_monitor_resume(mosh_at_cmd_mode_handler);
```
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:336]: removed duplicate

File: samples/nrf9160/modem_shell/src/at/at_cmd.c (hunk 7 above)
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/at/at_cmd.c:79]: fixed

File: samples/nrf9160/modem_shell/src/at/at_cmd_mode.c [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+send:
+
+	k_mutex_lock(&at_buf_mutex, K_FOREVER);
+	at_buf[at_cmd_len] = '\0'; /* Terminate the command string */
+
+	inside_quotes = false;
+	at_cmd_len = 0;
+	k_mutex_unlock(&at_buf_mutex);
+
+	/* Check for the presence of one printable non-whitespace character */
```
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/at/at_cmd_mode.c:213]: Removed whole block. Let it fail when parsing at cmd.

File: samples/nrf9160/modem_shell/src/at/at_cmd.c [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+#endif
+		char ch = *(at_cmd + i);
+		/* With parameter, SET TEST, "="; READ, "?" */
+		return ((ch == '=') || (ch == '?'));
+	}
+
+	return true;
+}
+
+int at_cmd_mosh_parse(const char *at_cmd)
```
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/at/at_cmd.c:67]: Agree, renamed.

File: samples/nrf9160/modem_shell/src/at/at_cmd.c [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+static struct mosh_at_cmd {
+	char *string;
+	mosh_at_handler_t handler;
+} mosh_at_cmd_list[] = {
+	{"AT+NPING", at_cmd_ping_handle},
+};
+
+#define MOSH_AT_CMD_MAX_PARAM 8
+
+static bool at_cmd_mode_util_mosh_cmd_casecmp(const char *at_cmd, const char *mosh_cmd)
```
[the remark this answers is not in the document]
Comment (jhirsi) [reply samples/nrf9160/modem_shell/src/at/at_cmd.c]: Agree. Renamed.

**Pull request merged**